coderd/httpmw/provisionerdaemon.go

```diff
@@ -0,0 +1,69 @@
+package httpmw
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/google/uuid"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/codersdk"
+)
+
+type provisionerDaemonContextKey struct{}
+
+// ProvisionerDaemon returns the daemon from the ExtractProvisionerDaemon handler.
+func ProvisionerDaemon(r *http.Request) database.ProvisionerDaemon {
+	user, ok := r.Context().Value(provisionerDaemonContextKey{}).(database.ProvisionerDaemon)
+	if !ok {
+		panic("developer error: provisioner daemon middleware not provided")
+	}
+	return user
+}
+
+// ExtractWorkspaceAgent requires authentication using a valid provisioner token.
+func ExtractProvisionerDaemon(db database.Store) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			cookie, err := r.Cookie(codersdk.SessionTokenKey)
+			if err != nil {
+				httpapi.Write(rw, http.StatusUnauthorized, codersdk.Response{
+					Message: fmt.Sprintf("Cookie %q must be provided.", codersdk.SessionTokenKey),
+				})
+				return
+			}
+			token, err := uuid.Parse(cookie.Value)
+			if err != nil {
+				httpapi.Write(rw, http.StatusUnauthorized, codersdk.Response{
```

Comment:
I'd consider this a bad request, but perhaps there's a reason it's unauthorized?

```suggestion
				httpapi.Write(rw, http.StatusBadRequest, codersdk.Response{
```

Reply:
This is mostly just for consistency with `ExtractWorkspaceAgent`, but I think a 401 error is reasonable here.

From a client's point of view, a token is an opaque string, and our implementation happens to generate tokens that look like UUIDs. Since we return a 401 error if the cookie is a valid UUID but isn't a token that exists in the DB, it makes sense to return the same error if it's not a valid UUID (and therefore can't be a valid token).